pennylane/argmap.py

```diff
@@ -0,0 +1,251 @@
+            to be parsed via ``dict(data)``.
+        single_arg (bool): Whether the ``ArgMap`` stores objects for a single argument.
+        single_object (bool): Whether the ``ArgMap`` stores only a single object.
+
+    An ``ArgMap`` acts as a dictionary with a restricted type of keys in the context of
+    a given ``QNode``. Each key takes the form ``tuple[int, tuple[int] or int]``.
+    The first ``int`` in the tuple describes the argument index of the ``QNode``,
+    while the second entry describes the index within that argument.
+    That is ``(2, (0, 3))`` describes the parameter in the ``0``th row and ``3``rd
+    column of the ``2``nd argument of a QNode. The second entry of the tuple may be an
+    ``int`` instead, indexing a one-dimensional array.
+
+    If the argument index (first entry) of a key is ``None``, this indicates that there is
+    only one argument in the QNode, and all keys of the ``ArgMap`` should have the form
+    ``(None, tuple[int] or int)``.
+    If the parameter index (second entry) of a key is ``None``, the respective argument
+    is a scalar and no other key with the same argument index can exist in the ``ArgMap``.
+    If both entries are ``None``, i.e. ``key=(None, None)``, the QNode only takes a single
+    scalar argument.
+
+    The main feature of an ``ArgMap`` is its flexible access: Users can specify only an
+    argument index or only a parameter index if this fully uniquely describes the
+    parameter.
+
+    **Example**
+    Suppose we have the following ``QNode`` instance:
+
+    .. code-block:: python
+
+        dev = qml.device('default.qubit', wires=2)
+        @qml.qnode(dev)
+        def circuit(arg1, arg2, arg3):
+            qml.Rot(arg1, wires=0)
+            qml.RX(arg2, wires=0)
+            qml.Rot(arg3, wires=1)
+            return qml.expval(qml.PauliY(0) @ qml.PauliY(1))
+
+    Then we might want to set up an ``ArgMap`` to store the generators of the rotations in
+    this circuit, depending on the parameter that controls them:
+
+    .. code-block:: python
+
+        generators = qml.ArgMap({
+            (0, 0): qml.PauliZ(0),
+            (0, 1): qml.PauliY(0),
+            (0, 2): qml.PauliZ(0),
+            1: qml.PauliX(0),
+            (1, 0): qml.PauliZ(1),
+            (1, 1): qml.PauliY(1),
+            (1, 2): qml.PauliZ(1),
+        })
+
+    Then the generators can be accessed by using the tuples ``(arg_index, par_index)`` as key
+    with ``par_index=None`` for the ``PauliX(0)`` generator. The latter may alternatively
+    be accessed directly via ``generators[1]``.
+    """
+
+    def __init__(self, data=None, single_arg=False, single_object=False, like=None):
+        if like is not None:
+            if not isinstance(like, ArgMap):
+                raise ArgMapError("Trying to inherit properties from non-ArgMap object.")
+            single_arg, single_object = like.single_arg, like.single_object
```

Comment:
Ah, sorry. This is because I added this functionality while checking out the implementation of `spectrum` and `reconstruct`. There, it was quite convenient to infer the `ArgMap` properties of the reconstruction `ArgMap` from the `ArgMap`s for the spectrum or the numbers of frequencies. Maybe it is a bit too much tayloring and simply invoking
```python
new_argmap = ArgMap(data, single_arg=old_argmap.single_arg, single_object=old_argmap.single_object)
```
also is fine?